### ISSUE TITLE
Centos 7 compilation failed

### ISSUE THREAD

**Issue by detailyang** [01org/hyperscan#56]
Hello. 
I'm failed to compile hyperscan on Centos7. It's the compiler error as the following:

```bash
make[2]: *** [CMakeFiles/hs.dir/src/nfa/dfa_build_strat.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /shared/test/deps/hyperscan-4.5.0/src/nfagraph/ng_holder.h:43:0,
                 from /shared/test/deps/hyperscan-4.5.0/src/nfagraph/ng_repeat.h:36,
                 from /shared/test/deps/hyperscan-4.5.0/src/nfa/castlecompile.h:39,
                 from /shared/test/deps/hyperscan-4.5.0/src/nfa/castlecompile.cpp:34:
/shared/test/deps/hyperscan-4.5.0/src/util/ue2_containers.h: In instantiation of ‘void ue2::flat_set<T, Compare, Allocator>::erase(ue2::flat_set<T, Compare, Allocator>::const_iterator) [with T = unsigned int; Compare = std::less<unsigned int>; Allocator = std::allocator<unsigned int>; ue2::flat_set<T, Compare, Allocator>::const_iterator = ue2::flat_detail::iter_wrapper<__gnu_cxx::__normal_iterator<const unsigned int*, std::vector<unsigned int> >, const unsigned int>; typename ue2::flat_detail::flat_base<T, Compare, Allocator>::storage_type::const_iterator = __gnu_cxx::__normal_iterator<const unsigned int*, std::vector<unsigned int> >]’:
/shared/test/deps/hyperscan-4.5.0/src/util/ue2_containers.h:295:21:   required from ‘void ue2::flat_set<T, Compare, Allocator>::erase(const key_type&) [with T = unsigned int; Compare = std::less<unsigned int>; Allocator = std::allocator<unsigned int>; ue2::flat_set<T, Compare, Allocator>::key_type = unsigned int]’
/shared/test/deps/hyperscan-4.5.0/src/nfa/castlecompile.cpp:726:27:   required from here
/shared/test/deps/hyperscan-4.5.0/src/util/ue2_containers.h:285:9: error: no matching function for call to ‘std::vector<unsigned int>::erase(__gnu_cxx::__normal_iterator<const unsigned int*, std::vector<unsigned int> >)’
         data().erase(pos.get());
```

My env is the following:

```bash
[root@localhost hyperscan-4.5.0]# cat /etc/redhat-release
CentOS Linux release 7.3.1611 (Core)

[root@localhost hyperscan-4.5.0]# gcc --version
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11)
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Any idea ?

**Comment by jviiret** [comment]
Thanks for the report. I have tracked this down and reproduced it locally.

This is because (a) you are using Boost 1.57, which is too early to provide `small_vector` as part of Boost.Container, and (b) you are using gcc-4.8, which has an incomplete implementation of the C++11 standard library. This causes issues with some compatibility code we have to fall back to using `std::vector` instead of `boost::small_vector` when it isn't available.

I'm working on a patch to fix this issue.

Right now, the best workaround is to use a version of Boost >= 1.58, or a newer GCC.

**Comment by detailyang** [comment]
@jviiret  Many thanks! After upgrading the Boost and GCC, It's okay.

**Comment by jviiret** [comment]
For anyone else who runs into this, there is a fix now in our develop branch, commit 09f5699.
